python/ray/autoscaler/updater.py

```diff
@@ -316,6 +329,25 @@ def run(self,
             port_forward=None,
             with_output=False)
 
+    def ensure_docker_installed(self):
+        try:
+            self.ssh_command_runner.run("command -v docker")
+            return
+        except Exception:
+            logger.info("Docker not installed, installing now")
+
+        install_commands = [
+            "curl -fsSL https://get.docker.com -o get-docker.sh",
+            "sudo sh get-docker.sh", "sudo usermod -aG docker $USER",
+            "sudo systemctl restart docker -f"
+        ]
+        for cmd in install_commands:
+            self.ssh_command_runner.run(cmd, ssh_options=SSHOptions())
+
+        logger.info("Docker install finished!")
+
+        # Restore initial SSH options.
+
```

Comment:
I'm not totally convinced this is a good idea. Shouldn't this just be done in the setup_commands or similar? I imagine that once we start auto-installing things this can be hard to maintain and a bit of a slippery slope.

Reply:
The main reason that I added this feature is because it is pretty much impossible to automate docker install with how we do SSH. For it to succeed the existing, persistent SSH has to end and timeout. The way for a user to do this is:
1. Write a YAML that installs docker
2. Run said YAML with `ray up <>.yaml`
--> `docker pull` will fail with **permission denied connecting to docker socket**
3. Edit YAML to exclude all docker install scripts
4. Wait ~10 seconds
5. Run `ray up <>.yaml`